stdlib/src/memory/maybe_uninitialized.mojo

```diff
@@ -38,14 +36,42 @@ struct UnsafeMaybeUninitialized[ElementType: CollectionElement](
         """The memory is now considered uninitialized."""
         self._array = __mlir_op.`kgen.undef`[_type = Self.type]()
 
+    @doc_private
     @always_inline
-    fn __init__(inout self, owned value: Self.ElementType):
+    fn __init__(inout self, *, other: Self):
+        """It is not possible to call this method.
+
+        Trying to call this method will fail early, at compile-time.
+        """
+        constrained[
+            False,
+            (
+                "You should never call the explicit copy constructor of"
+                " UnsafeMaybeUninitialized because it's ambiguous to copy"
+                " possibly uninitialized memory. Use"
+                " `UnsafeMaybeUninitialized.copy_from()` instead if you want to"
+                " trigger an explicit copy of the content of"
+                " UnsafeMaybeUninitialized. It has very specific semantics."
+            ),
+        ]()
+        self = Self()
```

Comment:
I think making this a compile time failure might prevent uses like `InlineArray[UnsafeMaybeUninitialized[T]]`, because that type would instantiate a call to this copy constructor.

The goal would be to avoid _dynamically_ doing a copy operation, but we need the compiler to accept code that "looks like" it could do a copy, because its up to the programming to make sure no actual copy gets attempted.